src/node_platform.cc

```diff
@@ -186,4 +200,10 @@ void TaskQueue<T>::Stop() {
   tasks_available_.Broadcast(scoped_lock);
 }
 
+template <class T>
+bool TaskQueue<T>::HasPending() {
+  Mutex::ScopedLock scoped_lock(lock_);
+  return !task_queue_.empty();
+}
```

Comment:
This method is intrinsically race-y.  Maybe change `FlushForegroundTasksInternal()` to return `true` when foreground tasks were dispatched and update the loop in `DrainBackgroundTasks()` to something like this?
```c++
while (FlushForegroundTasksInternal())
  background_tasks_.BlockingDrain();
```

Reply:
@bnoordhuis Fwiw, I don’t think the current code is actually race-y since, at least currently, only background tasks would enqueue new foreground tasks from another thread.

I like your suggestion, and d5bd7c55c19d0d4aee5a9511b9e6df1374d0644d should implement it correctly, but I think it doesn’t actually change the semantics here.

Reply:
If you mean it's not race-y now because it's always called from the right thread, yes, you're right.  What I mean is that the method itself is race-y by design.